src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs

```diff
@@ -312,7 +313,9 @@ private void BuildPreprocessorStateChangeMap()
             CSharpParseOptions? options = null,
             string path = "",
             Encoding? encoding = null,
+            // obsolete parameter -- unused
```

Comment:
We should document this in actual doc comments. We could also consider marking the method as obsolete, removing all the default parameters, and introducing a version that does not have these.

Reply:
So I guess that means essentially restoring one of the back compat overloads as the current one.

What should the right behavior be for the obsolete one? It'll legitimately ignore the isGenerated flag (since that's actually unused) but somebody passing diagnosticOptions will still get that passed through? Or @agocke did you want to preserve the behavior for somebody calling that?

Reply:
This method has two levels of "unused" -- actually unused versus just obsolete. :smile:

Reply:
IsGenerated was used before, but because there's no public API for generated files, I felt comfortable removing it.

As for changing this method -- I was not confident that obsoleting this and producing an overload was not a breaking change.

Reply:
Alright gave it a few shots and I agree with @agocke it's sure not obvious how to do this. Any call like

    Create(root, diagnosticOptions: something)

requires us to still have the "bad" methods to have optional parameters. But we'd also have to have a "good" method with the optional parameters as well. The compiler would then error on this being ambiguous:

    Create(root)

because it's possible to fill in parameters and hit either the bad method or good method.

Honestly I don't see any way to obsolete this without either the compiler having some tweak to overload resolution to treat Obsolete as a tiebreaker in the case above, or letting [Obsolete] be placed on a parameter which the compiler could flag if it's used as a named or positional parameter, but not if it's being filled in as an optional parameter. And yes I recognize that that previous sentence is "change overload resolution" or "change the framework since you can't put Obsolete on a parameter", neither of which seems like a good start. This may be worth a useful proposal somewhere but I suspect we don't want to block this PR on it's implementation. :smile:

I will at least update documentation and Obsolete the back compat overloads.

Reply:
@333fred and I chatted further and realized that if we do this, the only code we're breaking with a source-breaking change is the code that was trying to call methods passing in obsolete arguments. Since that's basically what we _want_ to do by obsoleting things, it's probably fine. The compatibility council agreed this was a good course of action, so I've updated the PR accordingly.

@333fred can you re-review? It seems like right now there's a bug with the public API checker which isn't flagging API removals which I think is impacting this; there's additions to the public API unshipped files but no requivalent `**REMOVE**` lines and that's worrisome.